File: libasynDiscordWebHook-Nukkit/src/main/java/com/ialdrich23xx/libasynwebhook/nukkit/discord/body/embed/EmbedManager.java
package com.ialdrich23xx.libasynwebhook.nukkit.discord.body.embed;

import com.ialdrich23xx.libasynwebhook.nukkit.Loader;
import com.ialdrich23xx.libasynwebhook.nukkit.discord.body.embed.base.Structure;
import com.ialdrich23xx.libasynwebhook.nukkit.discord.body.embed.components.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EmbedManager extends Structure {

    private String title;
    private String description;
    private Integer color;

    private Author author = null;
    private List<Field> fields = new ArrayList<>();
    private Footer footer = null;
    private Thumbnail thumbnail = null;
    private Image image = null;
    private Timestamp timestamp = null;

    public EmbedManager(String title, String description, Integer color) {
        this.title = title;
        this.description = description;
        this.color = color;
    }

    public static EmbedManager make(String title, String description, Integer color) {
        return new EmbedManager(title, description, color);
    }

    public static EmbedManager make(String title, String description) {
        return new EmbedManager(title, description, EmbedColors.Default);
    }

    public EmbedManager setTitle(String title) {
        this.title = title;

        return this;
    }

    public String getTitle() {
        return this.title;
    }

    public EmbedManager setDescription(String description) {
        this.description = description;

        return this;
    }

    public String getDescription() {
        return this.description;
    }

    public EmbedManager setColor(Integer color) {
        this.color = color;

        return this;
    }

    public Integer getColor() {
        return this.color;
    }

    public EmbedManager setAuthor(Author author) {
        if (!author.build()) {
            Loader.getInstance().getLogger().error("Author is invalid: " + author);
        } else this.author = author;

        return this;
    }

    public EmbedManager removeAuthor() {
        this.author = null;

        return this;
    }

    public Author getAuthor() {
        return this.author;
    }

    public EmbedManager addField(Field field) {
        if (!field.build()) {
            Loader.getInstance().getLogger().error("Field is invalid: " + field);
        } else this.fields.add(field);

        return this;
    }

    public EmbedManager resetFields() {
        this.fields = new ArrayList<>();

        return this;
    }

    public List<Field> getFields()
    {
        return this.fields;
    }

    public EmbedManager setFooter(Footer footer) {
        if (!footer.build()) {
            Loader.getInstance().getLogger().error("Footer is invalid: " + footer);
        } else this.footer = footer;

        return this;
    }

    public EmbedManager removeFooter() {
        this.footer = null;

        return this;
    }

    public Footer getFooter() {
        return this.footer;
    }

    public EmbedManager setThumbnail(Thumbnail thumbnail) {
        if (!thumbnail.build()) {
            Loader.getInstance().getLogger().error("Thumbnail is invalid: " + thumbnail);
        } else this.thumbnail = thumbnail;

        return this;
    }

    public EmbedManager removeThumbnail() {
        this.thumbnail = null;

        return this;
    }

    public Thumbnail getThumbnail() {
        return this.thumbnail;
    }

    public EmbedManager setImage(Image image) {
        if (!image.build()) {
            Loader.getInstance().getLogger().error("Image is invalid: " + image);
        } else this.image = image;

        return this;
    }

    public EmbedManager removeImage() {
        this.image = null;

        return this;
    }

    public Image getImage() {
        return this.image;
    }

    public EmbedManager setTimestamp(Timestamp timestamp) {
        if (!timestamp.build()) {
            Loader.getInstance().getLogger().error("Image is invalid: " + timestamp);
        } else this.timestamp = timestamp;

        return this;
    }

    public EmbedManager removeTimestamp() {
        this.timestamp = null;

        return this;
    }

    public Timestamp getTimestamp() {
        return this.timestamp;
    }

    @Override
    public Boolean build() {
        return !this.getTitle().isEmpty() && !this.getDescription().isEmpty() || !this.getFields().isEmpty();
    }

    @Override
    public Map<String, Object> toArray() {
        Map<String, Object> result = new HashMap<>();

        result.put("title", this.getTitle());
        result.put("description", this.getDescription());
        result.put("color", this.getColor());

        if (this.getAuthor() != null) result.put("author", this.getAuthor().toArray());
        if (this.getFooter() != null) result.put("footer", this.getFooter().toArray());
        if (this.getThumbnail() != null) result.put("thumbnail", this.getThumbnail().toArray());
        if (this.getImage() != null) result.put("image", this.getImage().toArray());
        if (this.getTimestamp() != null) result.put("timestamp", this.getTimestamp().getFormat().format(this.getTimestamp().getDate()));

        List<Map<String, Object>> fieldList = new ArrayList<>();

        this.getFields().forEach(field -> {
            fieldList.add(field.toArray());
        });

        if (!fieldList.isEmpty()) {
            result.put("fields", fieldList);
        }

        return result;
    }

    @Override
    public String toString() {
        return null;
    }
}

File: libasynDiscordWebHook-PM/src/ialdrich23xx/libasynwebhook/discord/body/embed/EmbedManager.php
<?php

declare(strict_types=1);

namespace ialdrich23xx\libasynwebhook\discord\body\embed;

use ialdrich23xx\libasynwebhook\discord\body\embed\base\Structure;
use ialdrich23xx\libasynwebhook\discord\body\embed\components\Author;
use ialdrich23xx\libasynwebhook\discord\body\embed\components\Field;
use ialdrich23xx\libasynwebhook\discord\body\embed\components\Footer;
use ialdrich23xx\libasynwebhook\discord\body\embed\components\Image;
use ialdrich23xx\libasynwebhook\discord\body\embed\components\Thumbnail;
use ialdrich23xx\libasynwebhook\discord\body\embed\components\Timestamp;
use ialdrich23xx\libasynwebhook\Loader;
use function is_null;
use function strlen;

class EmbedManager extends Structure
{
    private ?Author $author = null;

    /** @var Field[] $fields */
    private array $fields = [];

    private ?Footer $footer = null;

    private ?Thumbnail $thumbnail = null;

    private ?Image $image = null;

    private ?Timestamp $timestamp = null;

    public function __construct(
        private string $title,
        private string $description,
        private int $color = EmbedColors::Default
    ){}

    public static function make(string $title, string $description, int $color = EmbedColors::Default): self
    {
        return new self($title, $description, $color);
    }

    public function setTitle(string $title): self
    {
        $this->title = $title;

        return $this;
    }

    public function getTitle(): string
    {
        return $this->title;
    }

    public function setDescription(string $description): self
    {
        $this->description = $description;

        return $this;
    }

    public function getDescription(): string
    {
        return $this->description;
    }

    public function setColor(int $color): self
    {
        $this->color = $color;

        return $this;
    }

    public function getColor(): int
    {
        return $this->color;
    }

    public function setAuthor(Author $author): self
    {
        if (!$author->build()) {
            Loader::getInstance()->getPlugin()->getLogger()->error("Author is invalid: " . $author->toString());
        } else $this->author = $author;

        return $this;
    }

    public function removeAuthor(): self
    {
        $this->author = null;

        return $this;
    }

    public function getAuthor(): ?Author
    {
        return $this->author;
    }

    public function addField(Field $field): self
    {
        if (!$field->build()) {
            Loader::getInstance()->getPlugin()->getLogger()->error("Field is invalid: " . $field->toString());
        } else $this->fields[] = $field;

        return $this;
    }

    public function resetFields(): self
    {
        $this->fields = [];

        return $this;
    }

    /**
     * @return Field[]
     */
    public function getFields(): array
    {
        return $this->fields;
    }

    public function setFooter(Footer $footer): self
    {
        if (!$footer->build()) {
            Loader::getInstance()->getPlugin()->getLogger()->error("Footer is invalid: " . $footer->toString());
        } else $this->footer = $footer;

        return $this;
    }

    public function removeFooter(): self
    {
        $this->footer = null;

        return $this;
    }

    public function getFooter(): ?Footer
    {
        return $this->footer;
    }

    public function setThumbnail(Thumbnail $thumbnail): self
    {
        if (!$thumbnail->build()) {
            Loader::getInstance()->getPlugin()->getLogger()->error("Thumbnail is invalid: " . $thumbnail->toString());
        } else $this->thumbnail = $thumbnail;

        return $this;
    }

    public function removeThumbnail(): self
    {
        $this->thumbnail = null;

        return $this;
    }

    public function getThumbnail(): ?Thumbnail
    {
        return $this->thumbnail;
    }

    public function setImage(Image $image): self
    {
        if (!$image->build()) {
            Loader::getInstance()->getPlugin()->getLogger()->error("Image is invalid: " . $image->toString());
        } else $this->image = $image;

        return $this;
    }

    public function removeImage(): self
    {
        $this->image = null;

        return $this;
    }

    public function getImage(): ?Image
    {
        return $this->image;
    }

    public function setTimestamp(Timestamp $timestamp): self
    {
        if (!$timestamp->build()) {
            Loader::getInstance()->getPlugin()->getLogger()->error("Timestamp is invalid: " . $timestamp->toString());
        } else $this->timestamp = $timestamp;

        return $this;
    }

    public function removeTimestamp(): self
    {
        $this->timestamp = null;

        return $this;
    }

    public function getTimestamp(): ?Timestamp
    {
        return $this->timestamp;
    }

    public function build(): bool
    {
        return strlen($this->getTitle()) !== 0 && strlen($this->getDescription()) !== 0 || !empty($this->getFields());
    }

    public function toArray(): array
    {
        $result = [
            "title" => $this->getTitle(),
            "description" => $this->getDescription(),
            "color" => $this->getColor()
        ];

        if (!is_null($this->getAuthor())) $result["author"] = $this->getAuthor()->toArray();
        if (!is_null($this->getFooter())) $result["footer"] = $this->getFooter()->toArray();
        if (!is_null($this->getThumbnail())) $result["thumbnail"] = $this->getThumbnail()->toArray();
        if (!is_null($this->getImage())) $result["image"] = $this->getImage()->toArray();
        if (!is_null($this->getTimestamp())) $result["timestamp"] = $this->getTimestamp()->getData()->format(Timestamp::FORMAT);

        if (!empty($this->getFields())) {
            foreach ($this->getFields() as $field) {
                $result["fields"][] = $field->toArray();
            }
        }

        return $result;
    }

    public function toString(): string
    {
        return "EmbedManager(title=" . $this->getTitle() . ",description=" . $this->getDescription() . ",color=" . $this->getColor() .
            ",author=" . $this->getAuthor()?->toString() . ",footer=" . $this->getFooter()?->toString() . ",thumbnail=" . $this->getThumbnail()?->toString() .
            ",image=" . $this->getImage()?->toString() . ",timestamp=" . $this->getTimestamp()?->toString() . ")";
    }
}